### ISSUE TITLE
Enumerating package folders for .Net Core project

### ISSUE THREAD

**Issue by paveliak** [dotnet/sdk#11773]
Hi

I need to enumerate all package folders for a project basing on the `project.assets.json` file. The purpose is to extract licensing information about all project dependencies.

I couldn't find schema definition for that file but looking at the typical contents there seems to be two options. Could you help me understanding which of them is the preferred one:
- Use `project->restore->packagesPath` property and combine it with `project->restore->fallbackFolder`
```
"project": {
      "packagesPath": "/Users/someuser/.nuget/packages/",
      "projectStyle": "PackageReference",
      "fallbackFolders": [
        "/Users/someuser/MyProject/.dotnet/sdk/NuGetFallbackFolder"
      ],
```
- Use `packageFolders` property
```
"packageFolders": {
    "/Users/someuser/.nuget/packages/": {},
    "/Users/soimuser/MyProject/.dotnet/sdk/NuGetFallbackFolder": {}
  },
```

Thank you,
Pavel

**Comment by zarenner** [comment]
@nkolev92 any chance you're able to answer this?

**Comment by nkolev92** [comment]
The NuGet.g.props defines these in an MSBuild property, NuGetPackageFolders.

For example looking in my NuGet.Client projects nuget.g.props, I see:

```xml
    <NuGetPackageFolders Condition=" '$(NuGetPackageFolders)' == '' ">C:\Users\nikolev\.nuget\packages\;C:\Program Files\dotnet\sdk\NuGetFallbackFolder</NuGetPackageFolders>
```

Does that help/answer the question?

**Comment by zarenner** [comment]
@nkolev92 Not quite - in our scenario, we're only able to parse the `project.assets.json` (we cannot parse MSBuild). To locate a package on disk, we're trying to determine if it's more appropriate to search:
1. In order, the folders specified by `packageFolders`
or 
2. In order, the folders specified by both `project->packagesPath` and `project->fallbackFolders`

For example, a potential answer might be "you shouldn't use `packageFolders` since it's an object and the keys aren't guaranteed to be ordered", or alternatively "using only `packagesPath`+`fallbackFolders` will miss X that `packageFolders` would have". Or perhaps they're equivalent and doesn't matter which we use - we haven't seen a case where they differ, hence the question.

In short, which does NuGet itself actually use to resolve the package location?

**Comment by nkolev92** [comment]
I think you need to do the 1st one. 
I think that's what the SDK does. Basically the SDK uses: https://github.com/dotnet/sdk/blob/8288f55cf35fb642bf2838bd11b6e17d884f019f/src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs#L652
 which in turn uses https://github.com/dotnet/sdk/blob/8288f55cf35fb642bf2838bd11b6e17d884f019f/src/Tasks/Microsoft.NET.Build.Tasks/NuGetPackageResolver.cs.

The priority order that matter is global packages folder, then fallback folders.

**Comment by nkolev92** [comment]
How are you reading the assets file btw?

**Comment by paveliak** [comment]
Thank you @nkolev92. We use standard Ruby JSON parser to read it after restoring the projects with `dotnet restore`.

**Comment by dsplaisted** [comment]
If you were to look at the package folders on disk, you would typically be looking at all the packages that had been restored for all projects on that machine, rather than just the packages that the current project is using.
